### DISCUSSION TITLE
Rotation matrix

### DISCUSSION THREAD

**Discussion by pshen20127** [idaholab/moose#23445] [Q&A Modules: Solid mechanics] [answered]
Hello All

Do you know where in the codes is the place calculating the rotation matrix based on the input euler angles? which rotation convention moose uses? Does moose read in euler angles in radian or degree?

Thanks
pshen

**Comment by GiudGiud** [comment]
Hello

This file and example seems to have euler angles in degrees
https://github.com/idaholab/moose/blob/next/modules/combined/test/tests/grain_texture/grn_3_rand_2D.tex

You can have a look at this code for the rotation convention (though this one is the backward way, from the matrix)
https://mooseframework.inl.gov/docs/doxygen/modules/classComputeUpdatedEulerAngle.html

This class might be informative
https://mooseframework.inl.gov/docs/doxygen/modules/classEulerAngles.html

Guillaume

**Reply by YaqiWang** [reply]
`TransformGenerator` uses Euler angles as well for rotation. It calls a libMesh rotation function.

**Comment by sapitts** [comment, marked as answer]
Hi PShen,

I assume you are asking about the rotation matrix calculation for crystal plasticity, which uses the passive rotation rather than the active rotation MOOSE uses elsewhere (to better mimic experiments). 

That passive rotation used in the crystal plasticity codes is described in the documentation page here: https://mooseframework.inl.gov/source/materials/crystal_plasticity/ComputeElasticityTensorCP.html . Euler angle values are expected in degrees. The original lattice rotation is stored within the member variable `_crysrot[_qp]`.

The actual rotation matrix is computed from the Euler angles with the [RotationTensor class](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/utils/RotationTensor.C), and the updated rotation matrix is computed with the ComputeMultipleCrystalPlasticityStress class here: https://github.com/idaholab/moose/blob/d2db894c0a11cff9d35a7e62e6916a1ca60b95e1/modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C#L309-L312
The [ComputeUpdatedEulerAngle](https://mooseframework.inl.gov/source/materials/crystal_plasticity/ComputeUpdatedEulerAngle.html) class Guillaume recommended can be used to read and output the update crystal orientations.

Hope this helps,
Stephanie
